### PR TITLE
Fixed bug: In `Draft.get_clone_base` function.

### PR DESCRIPTION
**Pull request by amrit3701**
**Bug detail:** When we have clone object that is clone of clone another object, then in export IFC file object are displaced because `Draft.get_clone_base` not return correct base object.

```python
>>> Draft.get_clone_base(App.ActiveDocument.Clone).Label
'Base'
>>> Draft.get_clone_base(App.ActiveDocument.Clone001).Label
'CloneOfBase'  # Should also return `Base`

# After fix:
>>> Draft.get_clone_base(App.ActiveDocument.Clone).Label
'Base'
>>> Draft.get_clone_base(App.ActiveDocument.Clone001).Label
'Base'
```

### FreeCAD model
image

### Before bug fix
image

### After bug fix
image

[bug.zip](https://github.com/FreeCAD/FreeCAD/files/5262407/bug.zip)

### CODE REVIEW

**Comment by vocx-fc** [comment]
This code already handles recursive searching of the base object, however, it assumes the object that is cloned is placed in an Arch Component (this is any Arch Wall, Arch Structure, Arch Equipment, etc.).
```py
if hasattr(obj, "CloneOf"):
    if obj.CloneOf:
        return get_clone_base(obj.CloneOf)  # recursive

if get_type(obj) == "Clone" and obj.Objects:  # non-recursive
    return obj.Objects[0]
```

The first branch of this searches for the `"CloneOf"` property, which all Arch Components have. So, if you have a clone of an Arch Component, `get_clone_base` recursively searches for the base.

In the second branch, the object does not have a `"CloneOf"` property, meaning that the object is not an Arch Component; it's a regular object that was cloned. In this case it only returns the first element of the `Objects` list, meaning its immediate parent. It does not perform a recursive search.

So, what you are doing is forcing the function to always be recursive; this could impact the code everywhere that this function is used. However, I did a quick grep and the only place where this function seems to be used is in `exportIFC.py`. So the question is, do we always want this function to behave recursively? Then the change is fine; however, if we only want it to be recursive some times, then maybe we need to add a different parameter.
```py
def get_clone_base(obj, strict=False, recursive=True):

    if hasattr(obj, "CloneOf") and obj.CloneOf:
        if recursive:
            return get_clone_base(obj.CloneOf)
        return obj.ConeOf

    if get_type(obj) == "Clone" and obj.Objects:
        if recursive:
            return get_clone_base(obj.Objects[0])
        return obj.Objects[0]
```

Basically, the code works correctly at the moment, as long as objects are encapsulated in Arch Components. But if the user is not going to do that, then we need to adjust the code accordingly.

**Comment by amrit3701** [comment]
> This code already handles recursive searching of the base object, however, it assumes the object that is cloned is placed in an Arch Component (this is any Arch Wall, Arch Structure, Arch Equipment, etc.).
> 
> ```python
> if hasattr(obj, "CloneOf"):
>     if obj.CloneOf:
>         return get_clone_base(obj.CloneOf)  # recursive
> 
> if get_type(obj) == "Clone" and obj.Objects:
>     return obj.Objects[0]  # non-recursive
> ```
> 
> The first branch of this searches for the `"CloneOf"` property, which all Arch Components have. So, if you have a clone of an Arch Component, `get_clone_base` recursively searches for the base.
> 
> In the second branch, the object does not have a `"CloneOf"` property, meaning that the object is not an Arch Component; it's a regular object that was cloned. In this case it only returns the first element of the `Objects` list, meaning its immediate parent. It does not perform a recursive search.
> 
> So, what you are doing is not really "fixing a bug", rather you are changing the behavior of the function, you are forcing the function to always be recursive; this could impact the code everywhere that this function is used.
> 
> However, I did a quick grep and the only place where this function seems to be used is in `Mod/Arch/exportIFC.py`. So the question is, do we always want this function to behave recursively? Then the change is fine; however, if we only want it to be recursive some times, then maybe we need to add a different parameter.

Yes, this function is only used by `exportIFC` function. I always call recursively here because it's a normal behaviour how `draftobjects.clone.Clone` work. 

### Let's first understand how `draftobjects.clone.Clone` works:

Suppose, I create a sketch at (100, 0, 0) point and then pad it (let name of object is **Base**), so this **Base** object placed at (100, 0, 0) but `Base.Placement.Base` is (0, 0, 0). Now, I create a clone object (name **CloneOfBase**) on **Base** object, this **CloneOfBase** object placed exactly at the same point and taking **Base** object shape as origin point and when we set placement to (50, 0, 0) then this placement is from **Base** not from global origin.

Now, I again create another clone object (name **CloneOfClone**) on **CloneOfBase** object and `CloneOfClone.Placement.Base` is (50, 0, 0). If we change placement to (0, 0, 0), then this **CloneOfClone** placed exactly same at **Base** object. So, internally this object also taking **Base** object as base shape.

As, this check `get_type(obj) == "Clone" and obj.Objects:` is for `draftobjects.clone.Clone` object and that's why I always search base object recursively.

I don't think here we need to additional `recursive` argument.

**Comment by amrit3701** [comment]
@vocx-fc Added recursive argument as you suggested. 

https://github.com/FreeCAD/FreeCAD/pull/3887/commits/a95176eaac5542128ddb1304338d91033ca9078b
